File: Lodjinha/app/src/main/java/br/com/bsavoini/lodjinha/adapters/CategoriesAdapter.java
package br.com.bsavoini.lodjinha.adapters;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import br.com.bsavoini.lodjinha.R;
import br.com.bsavoini.lodjinha.api.model.CategoryModel;
import br.com.bsavoini.lodjinha.home.interfaces.CategoryClickCallback;
import com.squareup.picasso.Picasso;


import java.util.List;

public class CategoriesAdapter extends RecyclerView.Adapter<CategoriesAdapter.CategoryViewHolder> {
    private List<CategoryModel> categoriesArr;
    private Picasso picassoInstance;
    private CategoryClickCallback categoryClickCallback;

    public CategoriesAdapter(List<CategoryModel> categoriesArr, CategoryClickCallback categoryClickCallback) {
        this.categoriesArr = categoriesArr;
        this.categoryClickCallback = categoryClickCallback;
    }

    @NonNull
    @Override
    public CategoryViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        picassoInstance = Picasso.with(viewGroup.getContext());
        View view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.item_category, viewGroup, false);
        return new CategoryViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull CategoryViewHolder viewHolder, int i) {
        CategoryModel categoryModel = categoriesArr.get(viewHolder.getAdapterPosition());
        viewHolder.categoryTxt.setText(categoryModel.getDescription());

        picassoInstance.load(categoryModel.getImageURL())
                .placeholder(R.drawable.logo_navbar)
                .into(viewHolder.categoryImg);

    }

    @Override
    public int getItemCount() {
        return categoriesArr.size();
    }

    class CategoryViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        private TextView categoryTxt;
        private ImageView categoryImg;

        CategoryViewHolder(@NonNull View itemView) {
            super(itemView);
            categoryTxt = itemView.findViewById(R.id.txt_category);
            categoryImg = itemView.findViewById(R.id.img_category);
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            CategoryModel categoryModel = categoriesArr.get(getAdapterPosition());
            String categoryName = categoryModel.getDescription();
            int categoryId = categoryModel.getId();
            categoryClickCallback.onClickCategory(categoryName, categoryId);
        }
    }
}

File: app/src/main/java/marcus/com/br/b2wtest/ui/main/home/CategoriesAdapter.kt
package marcus.com.br.b2wtest.ui.main.home

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.item_category.view.*
import marcus.com.br.b2wtest.R
import marcus.com.br.b2wtest.model.data.CategoryData
import marcus.com.br.b2wtest.ui.BaseRecyclerAdapter

class CategoriesAdapter : BaseRecyclerAdapter<CategoryData>() {

    override fun onCreateViewHolder(parent: ViewGroup, type: Int): RecyclerView.ViewHolder {
        return ViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_category, parent, false))
    }

    override fun getItemCount(): Int {
        return recyclerList.size
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        (holder as ViewHolder).bind(recyclerList[position])
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        fun bind(categoryData: CategoryData) = with(itemView) {
            Picasso.get()
                .load(categoryData.urlImage)
                .placeholder(R.drawable.progress_animation)
                .error(R.drawable.vector_img_not_found)
                .into(itemCategoryImage)

            itemCategoryName.text = categoryData.description
        }
    }
}